File: training/csharp/src/De.Sots.CellarsAndCentaurs/Adapter/Persistence/CreatureRepository.cs
using System;

namespace De.Sots.CellarsAndCentaurs.Adapter.Persistence
{
    using Entity = De.Sots.CellarsAndCentaurs.Adapter.Persistence.CreatureEntity;

    public class CreatureRepository : Repository<Entity>
    {
        protected override Guid IdOf(Entity item) => item.Id;

        public Entity? FindByType(string typeName)
        {
            foreach (var entity in FindAll())
            {
                if (entity.Type.ToString() == typeName)
                {
                    return entity;
                }
            }
            return null;
        }
    }
}

File: sample-projects/swift/Sources/CellarsAndCentaurs/Adapter/Persistence/CreatureRepository.swift
import Foundation
import Logging

typealias Entity = CreatureEntity

final class CreatureRepository: Repository<Entity> {
    private let logger = Logger(label: "de.sots.cellarsandcentaurs.adapter.persistence.CreatureRepository")

    override func save(_ item: Entity) {
        logger.debug("saving entity \(item.id)")
        super.save(item)
    }

    func findByType(_ typeName: String) -> [Entity] {
        findAll().filter { $0.typeName == typeName }
    }
}
